### PR TITLE
Move gather_nd/scatter/scatter_nd_add op to the phi library

### PR DESCRIPTION
**Pull request by sneaxiy**
### PR types
Others

### PR changes
OPs

### Describe
Move `gather_nd/scatter/scatter_nd_add` op to the phi library.

### CODE REVIEW

File: paddle/phi/infermeta/ternary.cc [hunk 1]
```diff
@@ -89,4 +89,142 @@ void AddmmInferMeta(const MetaTensor& input,
   out->set_dtype(input.dtype());
 }
 
+void GatherNdGradInferMeta(const MetaTensor& x,
```
Comment (chenwhql) [inline paddle/phi/infermeta/ternary.cc]: grad的InferMeta推理不用，所以要做个代码隔离，我们统一放到backward.h/cc中了

File: paddle/phi/infermeta/ternary.cc [hunk 2]
```diff
@@ -89,4 +89,142 @@ void AddmmInferMeta(const MetaTensor& input,
...
+                           const MetaTensor& index,
+                           const MetaTensor& out_grad,
+                           MetaTensor* x_grad) {
+  x_grad->set_dims(x.dims());
```
Comment (chenwhql) [inline paddle/phi/infermeta/ternary.cc]: 如果可以的话，顺便set下dtype，原来的InferDtype是在kernel里mutable_data中做的，我们这次计划尽可能放到前置infer操作中，后续Alloc时对tensor dtype的set操作要去掉

File: paddle/phi/infermeta/ternary.cc (hunk 1 above)
Comment (sneaxiy) [reply paddle/phi/infermeta/ternary.cc]: Done.

File: paddle/phi/infermeta/ternary.cc (hunk 2 above)
Comment (sneaxiy) [reply paddle/phi/infermeta/ternary.cc]: Done.

**Review by chenwhql — APPROVED** [review]
LGTM

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Pull request merged**